# DL_Exercise 3: Group C

Group Members in MECE

In [4]:
import pandas as pd

# Define data for the MECE table
Group_C = {
    "Members": ["Aaryan Divedi", "Chandra shiva Rachavelpula","Dhvani Manishbhai Bhavani","Jaskaran Singh","Manan Rajeshbhai Kakadiya","Manjot Singh","Melvin Raju","Yamini Guduru"],
    "Contributions": ["Data Insights and Summary", "model evaluation insights and final results", "Report on Code quality, process and problem-solution","presentation", "Data Visualization","Feature Engineering","CNN Model Building, Compilation and Evaluation","Data gathering and Data preprocessing"]
}

# Create a pandas DataFrame
df = pd.DataFrame(Group_C)

# Display the MECE table
df

Members  \
0               Aaryan Divedi   
1  Chandra shiva Rachavelpula   
2   Dhvani Manishbhai Bhavani   
3              Jaskaran Singh   
4   Manan Rajeshbhai Kakadiya   
5                Manjot Singh   
6                 Melvin Raju   
7               Yamini Guduru   

                                       Contributions  
0                          Data Insights and Summary  
1        model evaluation insights and final results  
2  Report on Code quality, process and problem-so...  
3                                       presentation  
4                                 Data Visualization  
5                                Feature Engineering  
6     CNN Model Building, Compilation and Evaluation  
7              Data gathering and Data preprocessing

# Problem Statement and Explanation:


The Data Science team at True Brands, an e-commerce company, is tasked with developing a Convolutional Neural Network (CNN) model to accurately identify the brand of a given product based on its image. This model aims to enhance the company's product classification system, improve user search experience, and streamline inventory management by automatically categorizing products by brand from their images.






# Objective:

To predict which product image is more likely to be genuinely belonging to the brand provided. And
then to classify the product image into the brand that it belongs to.


# 1. Importing necessary libraries and packages:

In [ ]:
import pandas as pd
import os
import cv2
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelBinarizer


# 2. Importing Dataset:

In [ ]:
df=pd.read_csv('/content/brand_info.csv')

In [ ]:
directory_path = '/content/drive/MyDrive/images/images'

# Checking

images = os.listdir(directory_path)
print(images[0])
img_array = cv2.imread(os.path.join(directory_path,"39386.jpg"))
print(img_array)

18900.jpg
None


# 3. Understanding Data with number of columns and rows:

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15137 entries, 0 to 15136
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    15137 non-null  int64  
 1   ID            15137 non-null  int64  
 2   GenderType    15137 non-null  object 
 3   Type          15137 non-null  object 
 4   SubType       15137 non-null  object 
 5   Article       15137 non-null  object 
 6   PrimaryColor  15131 non-null  object 
 7   Seasonal      15136 non-null  object 
 8   Year          15136 non-null  float64
 9   Use           15133 non-null  object 
 10  Brand         15137 non-null  object 
dtypes: float64(1), int64(2), object(8)
memory usage: 1.3+ MB


In [ ]:
df.head()

Unnamed: 0     ID GenderType         Type     SubType       Article  \
0           1  39386        Men      Apparel  Bottomwear         Jeans   
1           2  59263      Women  Accessories     Watches       Watches   
2           4  53759        Men      Apparel     Topwear       Tshirts   
3           8  29114        Men  Accessories       Socks         Socks   
4          10   9204        Men     Footwear       Shoes  Casual Shoes   

  PrimaryColor Seasonal    Year     Use          Brand  
0         Blue   Summer  2012.0  Casual  Peter England  
1       Silver   Winter  2016.0  Casual          Titan  
2         Grey   Summer  2012.0  Casual           Puma  
3    Navy Blue   Summer  2012.0  Casual           Puma  
4        Black   Summer  2011.0  Casual           Puma

In [ ]:
df.describe()

Unnamed: 0            ID          Year
count  15137.000000  15137.000000  15136.000000
mean   22049.160402  24553.882672   2012.158496
std    12740.003039  15392.364801      1.778051
min        1.000000   1163.000000   2008.000000
25%    11097.000000  11093.000000   2011.000000
50%    22119.000000  22486.000000   2012.000000
75%    33066.000000  37185.000000   2012.000000
max    44117.000000  59838.000000   2018.000000

In [ ]:
df.columns

Index(['Unnamed: 0', 'ID', 'GenderType', 'Type', 'SubType', 'Article',
       'PrimaryColor', 'Seasonal', 'Year', 'Use', 'Brand'],
      dtype='object')

# 4. Finding Number of images:

In [ ]:
df.shape

(15137, 11)

In [ ]:
images = os.listdir(directory_path)
id_list = []
for file in images:
  img = file.split(".")[0]
  id_list.append(img)
print(len(id_list))

8690


In [ ]:
#  Delete rows without images found in folders

column_name = 'ID'
c = 0
print(df.shape)
# Iterate over each row in the specified column
for value in df[column_name]:
    if str(value) not in id_list:
        c = c + 1
        df = df[df[column_name] != value]
        # Add the DataFrame to the list
print(c)

print(df.shape)

(15137, 11)
11499
(3638, 11)


In [ ]:
df.head()

Unnamed: 0     ID GenderType         Type     SubType      Article  \
5           12  18653        Men     Footwear  Flip Flops   Flip Flops   
8           21  12967        Men     Footwear      Sandal      Sandals   
9           23  13089        Men      Apparel     Topwear  Sweatshirts   
13          40  18839        Men  Accessories       Belts        Belts   
16          46  18005        Men      Apparel  Bottomwear       Shorts   

   PrimaryColor Seasonal    Year     Use          Brand  
5         Black     Fall  2011.0  Casual           Fila  
8         Black     Fall  2011.0  Casual         Adidas  
9          Grey     Fall  2011.0  Sports         Adidas  
13        Black     Fall  2011.0  Formal  Peter England  
16        Black   Summer  2011.0  Sports           Puma

In [ ]:
df.shape

(3638, 11)

In [ ]:
df.info()
df['label'] = "real"


<class 'pandas.core.frame.DataFrame'>
Index: 3638 entries, 5 to 15134
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    3638 non-null   int64  
 1   ID            3638 non-null   int64  
 2   GenderType    3638 non-null   object 
 3   Type          3638 non-null   object 
 4   SubType       3638 non-null   object 
 5   Article       3638 non-null   object 
 6   PrimaryColor  3635 non-null   object 
 7   Seasonal      3638 non-null   object 
 8   Year          3638 non-null   float64
 9   Use           3637 non-null   object 
 10  Brand         3638 non-null   object 
dtypes: float64(1), int64(2), object(8)
memory usage: 341.1+ KB


In [ ]:
df['label'].unique()

array(['real'], dtype=object)

In [ ]:
df.columns

Index(['Unnamed: 0', 'ID', 'GenderType', 'Type', 'SubType', 'Article',
       'PrimaryColor', 'Seasonal', 'Year', 'Use', 'Brand', 'label'],
      dtype='object')

In [ ]:
image_labels = df.set_index('ID')['label'].to_dict()
image_labels

{18653: 'real',
 12967: 'real',
 13089: 'real',
 18839: 'real',
 18005: 'real',
 10866: 'real',
 15528: 'real',
 19123: 'real',
 13419: 'real',
 19311: 'real',
 14856: 'real',
 13841: 'real',
 12993: 'real',
 18495: 'real',
 19547: 'real',
 12732: 'real',
 10065: 'real',
 10859: 'real',
 10257: 'real',
 11385: 'real',
 18806: 'real',
 12500: 'real',
 15341: 'real',
 17416: 'real',
 12958: 'real',
 17624: 'real',
 13284: 'real',
 19578: 'real',
 11527: 'real',
 14869: 'real',
 15187: 'real',
 11143: 'real',
 11371: 'real',
 14603: 'real',
 13426: 'real',
 13846: 'real',
 12994: 'real',
 18492: 'real',
 19540: 'real',
 14851: 'real',
 19772: 'real',
 19316: 'real',
 19124: 'real',
 11349: 'real',
 10861: 'real',
 16198: 'real',
 12538: 'real',
 18002: 'real',
 12960: 'real',
 15379: 'real',
 11376: 'real',
 14604: 'real',
 19927: 'real',
 13421: 'real',
 16361: 'real',
 19329: 'real',
 11144: 'real',
 11520: 'real',
 11712: 'real',
 13879: 'real',
 17623: 'real',
 13283: 'real',
 10406: 

In [ ]:
row_count = df.shape[0]
print(row_count)

3638


In [ ]:

for file in id_list:

    try:
      if int(file) not in df['ID'].values:
        new_rows = {'Unnamed: 0': row_count, 'ID': int(file), 'GenderType': 'Female', 'Type': 'Dress', 'SubType': 'Casual', 'Article': 'Summer Dress', 'PrimaryColor': 'Red', 'Seasonal': 'Summer', 'Year': 2024, 'Use': 'Party', 'Brand': 'BrandZ','label':'fake'}

        # Convert new row into a DataFrame with specified column names
        new_row_df = pd.DataFrame([new_rows], columns=df.columns)

        # Concatenate the original DataFrame with the new row DataFrame
        df = pd.concat([df, new_row_df], ignore_index=True)
        row_count += 1
    except:
        continue



In [ ]:
df.shape

(8680, 12)

# 5. Identifyling Image: Fake or Real brand

In [ ]:
df['label'].unique()

array(['real', 'fake'], dtype=object)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8680 entries, 0 to 8679
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    8680 non-null   int64  
 1   ID            8680 non-null   int64  
 2   GenderType    8680 non-null   object 
 3   Type          8680 non-null   object 
 4   SubType       8680 non-null   object 
 5   Article       8680 non-null   object 
 6   PrimaryColor  8677 non-null   object 
 7   Seasonal      8680 non-null   object 
 8   Year          8680 non-null   float64
 9   Use           8679 non-null   object 
 10  Brand         8680 non-null   object 
 11  label         8680 non-null   object 
dtypes: float64(1), int64(2), object(9)
memory usage: 813.9+ KB


In [ ]:
df

Unnamed: 0     ID GenderType         Type     SubType       Article  \
0             12  18653        Men     Footwear  Flip Flops    Flip Flops   
1             21  12967        Men     Footwear      Sandal       Sandals   
2             23  13089        Men      Apparel     Topwear   Sweatshirts   
3             40  18839        Men  Accessories       Belts         Belts   
4             46  18005        Men      Apparel  Bottomwear        Shorts   
...          ...    ...        ...          ...         ...           ...   
8675        8675  10693     Female        Dress      Casual  Summer Dress   
8676        8676  10743     Female        Dress      Casual  Summer Dress   
8677        8677  10724     Female        Dress      Casual  Summer Dress   
8678        8678  10702     Female        Dress      Casual  Summer Dress   
8679        8679  10767     Female        Dress      Casual  Summer Dress   

     PrimaryColor Seasonal    Year     Use          Brand label  
0           Black     Fall  2011.0  Casual           Fila  real  
1           Black     Fall  2011.0  Casual         Adidas  real  
2            Grey     Fall  2011.0  Sports         Adidas  real  
3           Black     Fall  2011.0  Formal  Peter England  real  
4           Black   Summer  2011.0  Sports           Puma  real  
...           ...      ...     ...     ...            ...   ...  
8675          Red   Summer  2024.0   Party         BrandZ  fake  
8676          Red   Summer  2024.0   Party         BrandZ  fake  
8677          Red   Summer  2024.0   Party         BrandZ  fake  
8678          Red   Summer  2024.0   Party         BrandZ  fake  
8679          Red   Summer  2024.0   Party         BrandZ  fake  

[8680 rows x 12 columns]

Initializing Dataframe

In [ ]:
# Initialize the DataFrame to store image data and labels
cnn1 = pd.DataFrame(columns=['img_array', 'label'])

# 6. Creating a Dataframe with Image array and labels

In [ ]:
# Iterate through the IDs in the DataFrame
for id in df['ID']:
    # Construct the file name
    file_name = str(id) + ".jpg"

    print(os.path.join(directory_path, file_name))

    # Read the image
    try:
        img_array = cv2.imread(os.path.join(directory_path, file_name))
        if img_array is None:
            raise FileNotFoundError(f"File not found: {file_name}")
    except FileNotFoundError as e:
        print(e)
        continue

    # Check if the image is not None
    if img_array is not None:
        # Convert the image to grayscale
        gray_image = cv2.cvtColor(img_array, cv2.COLOR_BGR2GRAY)

        # Resize the image to 80x60 pixels
        resized_array = cv2.resize(gray_image, (80, 60))

        # Flatten the image array
        flat_array = resized_array.flatten()

        # Get the label for the image
        label_value = df[df['ID'] == id]['label'].values[0]

        # Create a new DataFrame row
        new_row = pd.DataFrame([[flat_array, label_value]], columns=["img_array", "label"])

        # Concatenate the new row to the original DataFrame
        cnn1 = pd.concat([cnn1, new_row], ignore_index=True)

# Print the DataFrame
print(cnn1)

Streaming output truncated to the last 5000 lines.
/content/drive/MyDrive/images/images/19034.jpg
/content/drive/MyDrive/images/images/19055.jpg
/content/drive/MyDrive/images/images/19074.jpg
/content/drive/MyDrive/images/images/19056.jpg
/content/drive/MyDrive/images/images/19021.jpg
/content/drive/MyDrive/images/images/19040.jpg
/content/drive/MyDrive/images/images/19036.jpg
/content/drive/MyDrive/images/images/19028.jpg
/content/drive/MyDrive/images/images/19064.jpg
/content/drive/MyDrive/images/images/19089.jpg
/content/drive/MyDrive/images/images/1905.jpg
/content/drive/MyDrive/images/images/19022.jpg
/content/drive/MyDrive/images/images/19020.jpg
/content/drive/MyDrive/images/images/19048.jpg
/content/drive/MyDrive/images/images/19079.jpg
/content/drive/MyDrive/images/images/19084.jpg
/content/drive/MyDrive/images/images/19061.jpg
/content/drive/MyDrive/images/images/19054.jpg
/content/drive/MyDrive/images/images/19067.jpg
/content/drive/MyDrive/images/images/19029.jpg
/content/d

# 7. **First CNN Model to deermine the REAL VS FAKE**


In [ ]:
# Convert the DataFrame columns to NumPy arrays
X = np.array(cnn1['img_array'].tolist())
y = np.array(cnn1['label'].tolist())

In [ ]:
X = X.reshape(-1, 60, 80, 1)

# 8. **Normalize the pixel value**

In [ ]:
X = X / 255.0

# 9. **Encode the labels**

In [ ]:
lb = LabelBinarizer()
y = lb.fit_transform(y)

# 10. **Split the data into training and testing sets**

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 11. Model Building

**Define the CNN model**

In [ ]:
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(60, 80, 1)),
    MaxPooling2D((2, 2)),
    Conv2D(64, (3, 3), activation='relu'),
    MaxPooling2D((2, 2)),
    Flatten(),
    Dense(128, activation='relu'),
    Dropout(0.5),
    Dense(1, activation='sigmoid')  # Use sigmoid for binary classification
])

**Compile the model**

In [ ]:
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# 12. **Train the model**

In [ ]:
model.fit(X_train, y_train, epochs=10, validation_data=(X_test, y_test))

Epoch 1/10
217/217 [==============================] - 56s 251ms/step - loss: 0.6755 - accuracy: 0.5912 - val_loss: 0.6399 - val_accuracy: 0.6215
Epoch 2/10
217/217 [==============================] - 59s 274ms/step - loss: 0.6252 - accuracy: 0.6418 - val_loss: 0.6132 - val_accuracy: 0.6492
Epoch 3/10
217/217 [==============================] - 49s 224ms/step - loss: 0.5833 - accuracy: 0.6734 - val_loss: 0.5734 - val_accuracy: 0.6901
Epoch 4/10
217/217 [==============================] - 52s 241ms/step - loss: 0.5483 - accuracy: 0.6976 - val_loss: 0.5612 - val_accuracy: 0.6947
Epoch 5/10
217/217 [==============================] - 49s 227ms/step - loss: 0.5203 - accuracy: 0.7285 - val_loss: 0.5442 - val_accuracy: 0.7172
Epoch 6/10
217/217 [==============================] - 53s 246ms/step - loss: 0.4923 - accuracy: 0.7470 - val_loss: 0.5467 - val_accuracy: 0.7154
Epoch 7/10
217/217 [==============================] - 50s 230ms/step - loss: 0.4674 - accuracy: 0.7628 - val_loss: 0.5354 - val_ac

# 13. **Evaluate the model**

In [ ]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print('Test accuracy:', test_acc)

55/55 [==============================] - 3s 53ms/step - loss: 0.5243 - accuracy: 0.7621
Test accuracy: 0.7620967626571655


# 14. **Predict on new data**

In [ ]:
predictions = model.predict(X_test)
predictions = (predictions > 0.5).astype(int)


55/55 [==============================] - 3s 53ms/step


# 15. Data Summary and Insights:

We built CNN model in which we got 76 % accuracy and can predict our new data.

Here, we can say our model is getting good accuracy for distinguish two types of images real and fake.